lib/pure/typetraits.nim

```diff
@@ -87,11 +87,11 @@ since (1, 1):
     # Note: `[]` currently gives: `Error: no generic parameters allowed for ...`
     type(default(T)[i])
 
-macro genericParams*(T: typedesc): untyped {.since: (1, 1).} =
-  ## return tuple of generic params for generic `T`
-  runnableExamples:
-    type Foo[T1, T2]=object
-    doAssert genericParams(Foo[float, string]) is (float, string)
+  type WrapStatic*[Val] = object
```

Comment:
Do not export this.

Reply:
that's needed for generic code though, so user can tell whether generic param is a static or a regular type (the runnableExamples would fail without exporting it)
the only other possibility I can think of is providing a new proc `proc isWrapStatic(T: typedesc): bool` to test for that but that's just more complex

Reply:
Hmm can't it produce ``static 1.0`` instead of ``WrapStatic[1.0]``?

Reply:
this can't work:
```
type Foo = (int, static[2])
Error: type expected, but got: 2
  type Foo = (int, static[2])
```
that's why WrapStatic is needed

Reply:
Ping @zah

Reply:
`static[2]` is not an appropriate notation. A more correct one will be `static[int](2)`. `static(2)` will be inferred to have the later type.

Now, could a tuple have a static type used as a field? I can imagine that supporting this would be quite a lot of work affecting many places in the compiler (e.g. when the field is accessed, it's considered a constant and no code should be generated for it in the back-ends, etc).

It should be usable like this:

```nim
type MyTyple = tuple(int, static int)

let x: MyTyple  = (readInt(), 10) # works
let y: MyTyple  = (10, readInt()) # error, `readInt()` is not a constant
```

But this tuple can be instantiated with any static integer. To truly bind the type of the tuple to a single possible value, we'll have to write something like this:

```nim
type TupleWithStaticInteger = (int, typeof(static(2)))

# To help with understanding the above, consider this tuple type:
type TupleWithoutStaticInteger = (int, typeof(2))
```

And that's the type that you'll have to produce in `genericParams` - pretty costly line of development, just to support this feature.

Reply:
Does that mean that you're fine with WrapStatic[T]?

Reply:
I'm not sure why `WrapStatic[T]` works. I guess somebody relaxed the rules for matching generic parameters in another PR. I've never intended for the following to work, but Nim-devel seems to be happy about it:

```nim
type
  Foo[T] = object

var x: Foo["test"]
```

For me, this signature of `Foo` should allow only type parameters. Even worse version of the code is the following:

```nim
type
  Foo[T] = object
    x: T

var x: Foo["test"]
echo x
```

It currently compiles and prints:

```
(x: "")
```

Reply:
but there's no typeclass for static (representing tyStatic) at the moment, eg:
`type WrapStatic[T:static] = object` doesn't work (while `type WrapStatic[T: static[float]]` does, but that's not useful here)
so we should support `type WrapStatic[T: static] = object`, regardless of whether  `[T]` should be allowed to match to a static (that's a debate to have elsewhere).
I think the way I wrote it is fine for today's nim, and once we add support for `[T: static]`, we'll simply update this code to use `type WrapStatic[T:static] = object`  instead
this can happen transparently without breaking any code that uses genericParams

Reply:
I guess I'm not fundamentally opposed to `WrapStatic`. It's helpful that it has less degrees of freedom compared to `tyStatic`. I would call it `StaticParam` and then name the payload param `value` as it is possible to access it with the dot operator.

But all of this won't be necessary if we give up on the notion that `genericParams` should return a tuple. We could have had another trait `genericParam(N): typedesc` that doesn't have these issues.

Reply:
>  I would call it StaticParam and then name the payload param value as it is possible to access it with the dot operator.

done

> But all of this won't be necessary if we give up on the notion that genericParams should return a tuple. We could have had another trait genericParam(N): typedesc that doesn't have these issues.

Note that this was my previous approach (https://github.com/nim-lang/Nim/pull/8554) but returning a tuple (which was your suggestion) is a better API, in particular it avoids needing 2 separate APIs (genericParam(N) + genericParamLen). 
That being said, I don't understand this argument, you'd have the exact same problem with `genericParam(N)` if the Nth generic param is static:

type Bar[N: static int, T] = object
type Bar3 = Bar[3, float]

type t = genericParams(Bar3, 1)
doAssert t is float
type t2 = genericParams(Bar3, 0) # won't work unless you also use `StaticParam`
(or you'd need yet another API to check whether the N'th param is static, eg
```nim
when genericParamsIsStatic(N):
  const value = genericParamsValue(N)
else:
  type t = genericParams(N)
```

what I have in this PR is arguably much simpler to use

Reply:
Yeah, I've gone through the same motions in my own thinking.

Perhaps one day [typedescs-as-values](https://github.com/nim-lang/Nim/commit/badb6c0f665e22bff77d91bb2f64f195110b9887) will have their revenge and we'll be able to make an alternative to `genericParams` that returns a real tuple value, not a type:

```nim
doAssert generamParams(Bar3) == (3, float)
```

Anyway, merging this is fine to me, but it would be nice if the funny business in #13529 is not postponed for later.